### PR TITLE
[Azure] Fixes `error: parameter virtualMachineScaleSetName cannot be empty`

### PR DESCRIPTION
**Pull request by michaelact**
Fixes https://github.com/prometheus/prometheus/issues/13643

## Problem
Cuplikan layar 2024-03-04 184059

### CODE REVIEW

File: discovery/azure/azure.go
Comment (machine424) [inline discovery/azure/azure.go:424]: It's worth checking why we ended up with "duplicated" code like this, maybe it's a bad merge. Maybe that merge broke other things...

Normally, https://github.com/prometheus/prometheus/commit/0704c7254876fefbdde85fecb8ad389106312af2 already brought the code to the state you requested.

Also it'd be great if we have a regression test for this.

Comment (machine424) [reply discovery/azure/azure.go:424]: Good catch by the way!

Comment (michaelact) [reply discovery/azure/azure.go:424]: > It's worth checking why we ended up with "duplicated" code like this, [...]

I'm not sure if we already have a tool to check for "duplicate code." If not, I recommend that we consider using a code quality scanner tool https://www.sonarsource.com/blog/manage-duplicated-code-with-sonar/ (open source code available at https://github.com/SonarSource/sonarqube).

> Normally, https://github.com/prometheus/prometheus/commit/0704c7254876fefbdde85fecb8ad389106312af2 already brought the code to the state you requested.

Do you want to revert that commit, or should I leave the pull request open until it gets merged? I encountered this issue actually 😅

> Also it'd be great if we have a regression test for this.

I think we could create a new issue for that and see if others can help with it.

Comment (machine424) [reply discovery/azure/azure.go:424]: I was trying to understand how we ended up with the code in main because https://github.com/prometheus/prometheus/commit/0704c7254876fefbdde85fecb8ad389106312af2 (that was already merged into main) already brought the code to the state of your PR.

I think the merge https://github.com/prometheus/prometheus/pull/13399 is the one that added that duplicated/stale code back by mistake.

I'll let the maintainers decide for the need of a regression test.

**Comment by beorn7** [comment]
The flaky Go test should go away after a rebase.

WRT test for merge failures: I don't think we can solve that general problem. Undetected merge failures can happen, as long as Git doesn't flag them and none of the unit tests breaks because of them.

**Comment by machine424** [comment]
> WRT test for merge failures: I don't think we can solve that general problem. Undetected merge failures can happen, as long as Git doesn't flag them and none of the unit tests breaks because of them.

But how about always going through `main` first? if we need to fix/change sth on an already cut release tag/branch, we make the change first on main, that way we can be sure that `Merge XXX release to main` merges like https://github.com/prometheus/prometheus/pull/13399 should not be touching any .go files but are only to update the changelog or other files. (if we go further and start building the changelog on main, we could even avoid the need to merge)

And yes, this is why I'd like to see a unit test for the change in this PR, to avoid regressions.

**Comment by beorn7** [comment]
Then we needed to cherry-pick all bug fixes. And have all the commits twice for the same fix.

Those kind of merge conflicts can also happen in the normal flow, I don't think that the "merge back to main" merges are special here.

**Comment by machine424** [comment]
I think it's better that the bug fix author assists merging the fix into main, and help the release shepherd (they're here for this :)) cherry-pick to the release branch, than the release shepherd having to do alone such blind "merge back to main" (I don't think fix authors review those PRs).

And this is just a proposition.

**Review by bboreham — APPROVED** [review]
Thanks for this, @michaelact.

I wrote a unit test at #13741 to check that this makes the problem go away. 

I am not familiar with Azure APIs, so it would be great if you could cast your eye over that PR.

**Pull request merged**
